### PR TITLE
Add support for the Waveshare ESP32-S3-Touch-LCD1.28 board

### PR DESCRIPTION
**Pull request by joelguittet**
The purpose of this Pull Request is to add support for the Waveshare ESP32-S3-Touch-LCD1.28 board.

### CODE REVIEW

File: drivers/input/input_cst816s.c
Comment (fabiobaltieri) [inline drivers/input/input_cst816s.c:18]: may be worthwile adding a short comment before these mentioning that the first comes from the official documentation and the second has been found on some devices but is not documented

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi
Comment (fabiobaltieri) [inline boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi:15]: nit:

```
                        pinmux = <I2C1_SDA_GPIO6>, <I2C1_SCL_GPIO7>;
```

no need to overindent, fits well within 80 columns

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi
Comment (fabiobaltieri) [inline boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi:26]: nit:

```
                        pinmux = <LEDC_CH0_GPIO2>, 
                                 <LEDC_CH1_GPIO5>,
                                 <LEDC_CH2_GPIO4>;
```

(not a hard rule but that's how it's commonly done)

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts
Comment (fabiobaltieri) [inline boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:54]: you can also use `PWM_HZ`, may make more sense if you are matching with a spec in the datasheet given in frequency

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig
Comment (fabiobaltieri) [inline boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig:10]: I think that normally subsystems options apart from the console one are left up to the application, @kartben?

**Review by fabiobaltieri — COMMENTED** [review]
cool board

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts
Comment (fabiobaltieri) [inline boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:9]: missed a dep?

Comment (fabiobaltieri) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:9]: Oh, https://github.com/zephyrproject-rtos/zephyr/pull/67012, I see that now.

File: drivers/input/input_cst816s.c
Comment (joelguittet) [reply drivers/input/input_cst816s.c:18]: If you are able to point me the official doc, good, but even that I'm not able to find it.

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi:26]: This is mixing tab and spaces, right ? So depending of the tab configuration in your editor, sometimes it's aligned and sometimes not (a reason for which I hate tab in source code).

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:54]: No specification here, this is a value I also seen on another board. I just copy/paste.

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig:10]: I'm curious of the strategy here. Also curious about why there is _defconfig and Kconfig.defconfig. Sometimes they overlap and I still not understand the usage of both files.

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:9]: Yes exactly, this is why some twistter tests are failing here

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi:26]: By the way, this can fit on 80 characters, so I will remove the \n

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts
Comment (fabiobaltieri) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:54]: ok, is 100Hz enough? no flicker at all? I see from https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-1.28/ESP32-S3-Touch-LCD-1.28-Sch.pdf that this drivers the LEDs directly, I'd suggest picking something faster, maybe 250Hz or something

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi
Comment (fabiobaltieri) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd-pinctrl.dtsi]: tabs + space is ok, space + tab or other mixes are not (and should be caught by the presubmit check), it does makes diff odd sometimes, this is just how things are

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig:10]: Instead of LVGL maybe "if DISPLAY" ? just in case somebody write a display application without LVGL support
Can be the following in Kconfig.defconfig:
```
config PWM
        default y if DISPLAY
```

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd.dts:54]: 100Hz seems enough yes, but 250Hz is fine as well.

File: boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig
Comment (fabiobaltieri) [reply boards/xtensa/waveshare_esp32s3_touch_lcd/waveshare_esp32s3_touch_lcd_defconfig]: yeah that works, note that I put a comment on https://github.com/zephyrproject-rtos/zephyr/pull/67012#discussion_r1444701046 about PWM usage on that driver, I think that the pwm backlight implementation could use a more generic implementation, let's follow up on that PR but the change will certainly reflect back on this one

**Comment by fabiobaltieri** [comment]
@joelguittet one more thing, you named the board `waveshare_esp32s3_touch_lcd` but when I see that Waveshare has two "waveshare esp32s3 touch lcd" boards, a 1.28'' and a 4.3'', could you rename the board to something like `waveshare_esp32s3_touch_lcd_1_28`? (We have a `buydisplay_2_8_tft_touch_arduino` shield already)

**Comment by kartben** [comment]
> @joelguittet one more thing, you named the board `waveshare_esp32s3_touch_lcd` but when I see that Waveshare has two "waveshare esp32s3 touch lcd" boards, a 1.28'' and a 4.3'', could you rename the board to something like `waveshare_esp32s3_touch_lcd_1_28`? (We have a `buydisplay_2_8_tft_touch_arduino` shield already)

It might make sense to keep the generic folder name though so that it can become the home for `waveshare_esp32s3_touch_lcd_4_3`, should it be added someday?

**Comment by fabiobaltieri** [comment]
> It might make sense to keep the generic folder name though so that it can become the home for `waveshare_esp32s3_touch_lcd_4_3`, should it be added someday?

They really don't have much in common https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-1.28/ESP32-S3-Touch-LCD-1.28-Sch.pdf https://files.waveshare.com/wiki/ESP32-S3-Touch-LCD-4.3/ESP32-S3-Touch-LCD-4.3-Sch.pdf, normally the board directory is for different revisions or variation with common pinout or stuff like that, but these are really two different boards, they barely use the same chip (one uses a module).

**Comment by kartben** [comment]
Ah! Sorry then :) Product naming is always funny. There I was, thinking only display size would be different based on the names 😅

**Comment by joelguittet** [comment]
Good point, yes I just checked too and they are very different. Will rename this one now.

File: samples/drivers/adc/boards/waveshare_esp32s3_touch_lcd_1_28.overlay
Comment (fabiobaltieri) [inline samples/drivers/adc/boards/waveshare_esp32s3_touch_lcd_1_28.overlay:9]: nit: leave it inline

**Review by fabiobaltieri — COMMENTED** [review]
looks good, guess we have to wait on the display pr

File: samples/drivers/adc/boards/waveshare_esp32s3_touch_lcd_1_28.overlay
Comment (joelguittet) [reply samples/drivers/adc/boards/waveshare_esp32s3_touch_lcd_1_28.overlay:9]: Yes will do

Comment (joelguittet) [reply samples/drivers/adc/boards/waveshare_esp32s3_touch_lcd_1_28.overlay]: Done

File: samples/subsys/display/lvgl/boards/waveshare_esp32s3_touch_lcd_1_28.conf
Comment (jfischer-no) [inline samples/subsys/display/lvgl/boards/waveshare_esp32s3_touch_lcd_1_28.conf:1]: Ohh, our minimal LVGL sample is not minimal anymore? What is the display resolution? Would you need same one-line-overlay for another LVGL sample?

**Comment by joelguittet** [comment]
@fabiobaltieri @kartben I see the tests for I2C is not passing but this look strange. Something related to the device tree. Have you seen this before ?

**Comment by fabiobaltieri** [comment]
No first time I see it, what's happening is that that specific test is defining a `vnd,i2c-device` node and defining a device for it but then this board is enabling `CONFIG_INPUT_CST816S`, which is selecting `CONFIG_I2C=y`, which in turn resulting in enabling `CONFIG_I2C_TEST=y` which builds `./drivers/i2c/i2c_test.c` which also defines a device for `vnd,i2c-device`, and that's how you end up with that handle defined twice. I think that the right thing to do would be to change that `select I2C` to `imply I2C`, but clearly the trend so far has been to exclude affected boards from the test, so for now I would suggest to just add this board with the others in `platform_exclude` in `tests/lib/devicetree/devices/testcase.yaml`.

Long term I suspect the right thing to do would be to just drop that list and instead put a `platform_allow: native_sim` but let's handle that separately.

**Comment by joelguittet** [comment]
@fabiobaltieri thanks for this detailed answer, will do that.

**Comment by fabiobaltieri** [comment]
Apologies for sitting on this, opened https://github.com/zephyrproject-rtos/zephyr/pull/68091 to restrict that test, if it gets merged first you could just drop the last patch.

**Comment by fabiobaltieri** [comment]
@joelguittet that test got filtered, you can drop the last patch, sorry for the noise

**Comment by joelguittet** [comment]
@fabiobaltieri PR updated here :)

File: samples/subsys/display/lvgl/src/main.c
Comment (fabiobaltieri) [inline samples/subsys/display/lvgl/src/main.c:159]: you can use `pwm_set_pulse_dt(&pwm_lcd0, pwm_lcd0.period)`, that way it picks the frequency from the tree, not that it matters in this case but that 250Hz should not be here

File: samples/subsys/display/lvgl/src/main.c
Comment (fabiobaltieri) [inline samples/subsys/display/lvgl/src/main.c:25]: I'd use a chosen rather than an alias, then it goes 1:1 with the display, something like this maybe?

```
		zephyr,display = &gc9a01;
		zephyr,display-backlight-pwm = &pwm_lcd0;
```

File: samples/subsys/display/lvgl/src/main.c
Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c:159]: I was really hopping to get a review on that because the 250Hz value was something we want in the sample. Thanks for indicate that, I have checked the pwm_* APIs but forgotten to just simply check for relevant members in the struct !!

File: samples/subsys/display/lvgl/src/main.c
Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c:25]: Well in this case everything should be converted to something like `zephyr,xxx` ? I based this proposal on leds/buttons that are simply something like `sw0 = &button0;`

By the way if we do that, then you can get a device with `DEVICE_DT_GET(DT_CHOSEN(zephyr_display_backlight_pwm))` I guess. I it really wanted to consider the backlight like a "device" ? It's not really one according to me...

File: boards/xtensa/waveshare_esp32s3_touch_lcd_1_28/Kconfig.defconfig
Comment (jfischer-no) [inline boards/xtensa/waveshare_esp32s3_touch_lcd_1_28/Kconfig.defconfig:28]: "Unless explicitly recommended otherwise by this section, leave peripherals and their drivers disabled by default." https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#general-recommendations

File: samples/subsys/display/lvgl/src/main.c
Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c]: Done

File: samples/subsys/display/lvgl/src/main.c
Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c:75]: Done

File: boards/xtensa/waveshare_esp32s3_touch_lcd_1_28/Kconfig.defconfig
Comment (joelguittet) [reply boards/xtensa/waveshare_esp32s3_touch_lcd_1_28/Kconfig.defconfig]: Sure, done!

File: samples/subsys/display/lvgl/src/main.c
Comment (fabiobaltieri) [reply samples/subsys/display/lvgl/src/main.c:25]: Not sure I get your point on "get a device" it's the same thing you'd just do `PWM_DT_SPEC_GET(DT_CHOSEN(zephyr_display_backlight_pwm));`, also I would use a separate define for the node so that you can use it in the various `DT_NODE_EXISTS`

```c
#define BACKLIGHT_PWM_NODE DT_CHOSEN(zephyr_display_backlight_pwm)
...
static const struct pwm_dt_spec pwm_lcd0 = PWM_DT_SPEC_GET(BACKLIGHT_PWM_NODE);
...
#if DT_NODE_EXISTS(BACKLIGHT_PWM_NODE)
...
etc
```

not sure about the alias vs chosen, they look interchangeable to me but I feel like this node should sit next to the display one. @jfischer-no any opinion? I don't feel strongly about it btw it just looks like it makes sense to me.

Comment (jfischer-no) [reply samples/subsys/display/lvgl/src/main.c:25]: In my opinion, this PR hides issues that should be fixed/discussed in a separate PR. A PR like "Add support for the foobar board" should only be about the board itself. Anything that fixes/touches samples/subsystems/drivers should be moved to a separate PR.

Comment (fabiobaltieri) [reply samples/subsys/display/lvgl/src/main.c:25]: Are you asking to split the sample PWM backlight out in its own PR? I don't disagree...

Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c:25]: I don t disagree as well. Do you want 2 parallel PR ? The name of the node is dependent for each other.
What about the first commits about the touchscreen ic id and the bug fix on the display driver ? Separated or not ?

Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c:25]: It will be also easier to review each topic separately for sure :-) initially my intention was just to have the board support here, but one by one I discovered issues by checking with the samples :-)

Comment (jfischer-no) [reply samples/subsys/display/lvgl/src/main.c:25]: @joelguittet Looks like your first three commits could be merged very quickly. Your board commit looks good. So yes, please split it into two or three PRs, whatever you prefer.

Comment (joelguittet) [reply samples/subsys/display/lvgl/src/main.c:25]: Good, will do ! After the fosdem this weekend :-) so should be done next week

Comment (fabiobaltieri) [reply samples/subsys/display/lvgl/src/main.c:25]: FYI RC1 is going to be cut next Monday I think so it's either now or after the release (Feb 25th).

**Comment by joelguittet** [comment]
This PR has now been separated as discussed with @fabiobaltieri and @jfischer-no and depends of:
- fix for the display driver at https://github.com/zephyrproject-rtos/zephyr/pull/68520
- additional cst816s ID at https://github.com/zephyrproject-rtos/zephyr/pull/68521

In parallel, in order to get the backlight of the screen ON, the LVGL sample has been modified at https://github.com/zephyrproject-rtos/zephyr/pull/68522 to support the PWM output.

**Review by fabiobaltieri — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]
Needs to be migrated to hwmv2

**Comment by joelguittet** [comment]
@nordicjm if my understanding is correct this mean moving on top of collab-hwm branch, then looking at existing boards already updated there, correct ? I have seen all the xtensa boards are currently in the legacy directory.

**Comment by nordicjm** [comment]
> @nordicjm if my understanding is correct this mean moving on top of collab-hwm branch, then looking at existing boards already updated there, correct ? I have seen all the xtensa boards are currently in the legacy directory.

Yes though espressif are not

**Comment by sylvioalves** [comment]
It still needs to be ported to hwmv2.

**Comment by joelguittet** [comment]
@sylvioalves true! I was waiting other esp32 board to be migrated to have an example and the common soc files I need here. It's now updated here!

The board `esp32s3_touch_lcd_1_28/esp32s3/procpu` has been tested with the following samples:
- samples/hello_world => works fine
- samples/drivers/adc/ to read the battery voltage (overlay is provided) => works fine
- samples/subsys/display/lvgl/ to check the display and the touchscreen controller => works fine

The board `esp32s3_touch_lcd_1_28/esp32s3/appcpu` has been tested compiled with the `samples/hello_world` but it's not booting. esp32s3_devkitm has the same behavior. I guess this is not due to this new waveshare board, but to be honest, I don't know the behavior of the appcpu board / how to test it... if somebody can guide me on this part ?

**Comment by nordicjm** [comment]
This needs to be targeted to `main` not the `collab-hwm` branch

**Comment by fabiobaltieri** [comment]
@joelguittet you can change the base of the existing PR, just click edit on the top and the base should become a dropbox and you can pick main there

**Comment by joelguittet** [comment]
@nordicjm @fabiobaltieri are you sure of your request ? If I target main then the PR will show all the commits of the collab hwm branch..... And if my PR is merged then it will merge everything in main including all the collab hwm commits.
Maybe you want me to target main, review only the last 3 commits and wait until collab hwm is merge before merging my PR ? If collab branch is merge and deleted GitHub should automatically retarget this PR to main so should not be blocking? (It s working like that on bitbucket but I m not sure on GitHub to be honest)

Note : I will fix the doc issue reported by the pipeline asap of course.

**Comment by nordicjm** [comment]
You need to rebase it properly so that your changes apply to main, not the collab-hwm branch, so this PR needs to contain just your commits that would clearly apply

**Comment by fabiobaltieri** [comment]
yeah well if you don't change the base it's going to be merged on the branch which is certainly not what you want either :-)

**Comment by joelguittet** [comment]
Oh my bad 😞 seems I m tired, just realized the collab hwm branch feature have merged to main branch too...
I was thinking it was not done yet and still a rework in progress
For sure now I will rebase on main !

**Comment by joelguittet** [comment]
Updated :)

File: boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts
Comment (nordicjm) [inline boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts:54]: Would remove and give to slots instead since swap using scratch is not recommended

File: boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_procpu.dts
Comment (nordicjm) [inline boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_procpu.dts:96]: as above

**Review by nordicjm — COMMENTED** [review]
Comments are optional

File: boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts
Comment (joelguittet) [reply boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts:54]: Agree but it's retrieved from the esp32s3_devkitm board from Espressif. I guess it's better to keep consistency across similar boards?

Comment (joelguittet) [reply boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts:54]: @nordicjm what is your opinion?

**Comment by joelguittet** [comment]
@nordicjm gentle ping for the above question, thanks!

File: boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts
Comment (nordicjm) [reply boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts:54]: @sylvioalves

Comment (joelguittet) [reply boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts:54]: @sylvioalves ? Your opinion about this ? What should be the default in zephyr for this kind of board ?

**Review by fabiobaltieri — DISMISSED** [review]

File: boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts
Comment (fabiobaltieri) [reply boards/waveshare/esp32s3_touch_lcd_1_28/esp32s3_touch_lcd_1_28_esp32s3_appcpu.dts:54]: @nordicjm hey since we are on the topic I'll add a question: do you think it would make sense to have few templates for in-tree board partitions maps? We have a lot of boards based on esp32, nrf52840 etc... and the flash partitioning is pretty much always the same, maybe we can have some dtsi and `#include` it in the board file? As much as I dislike mid-structure includes, it may be better than repeating this to infinity.

File: boards/waveshare/esp32s3_touch_lcd_1_28/doc/index.rst
Comment (joelguittet) [reply boards/waveshare/esp32s3_touch_lcd_1_28/doc/index.rst]: Nice catch! Done. And PR rebased on master!

**Review by fabiobaltieri — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Comment by joelguittet** [comment]
Thanks for the reviews here ! 🎉

**Pull request merged**
